wagtail/contrib/settings/views.py

```diff
@@ -82,118 +86,280 @@ def redirect_to_relevant_instance(request, app_name, model_name):
 
-        if request.method == "POST":
-            form = form_class(
-                request.POST, request.FILES, instance=instance, for_user=request.user
+        if site is not None:
+            return redirect(
+                "wagtailsettings:edit",
+                app_name=self.model._meta.app_label,
+                model_name=self.model._meta.model_name,
+                pk=site.pk,
             )
-
-            if form.is_valid():
-                with transaction.atomic():
-                    form.save()
-                    log(instance, "wagtail.edit")
-
-                messages.success(
-                    request,
-                    _("%(setting_type)s updated.")
-                    % {
-                        "setting_type": capfirst(setting_type_name),
-                        "instance": instance,
-                    },
-                )
-                return redirect("wagtailsettings:edit", app_name, model_name)
-            else:
-                messages.validation_error(
-                    request, _("The setting could not be saved due to errors."), form
-                )
         else:
-            form = form_class(instance=instance, for_user=request.user)
-
-        edit_handler = edit_handler.get_bound_panel(
-            instance=instance, request=request, form=form
-        )
+            return redirect(
+                "wagtailsettings:edit",
+                app_name=self.model._meta.app_label,
+                model_name=self.model._meta.model_name,
+                pk=self.object.pk,
+            )
 
-        media = form.media + edit_handler.media
+    def form_valid(self, form):
+        with transaction.atomic():
+            form.save()
+            log(self.object, "wagtail.edit")
+
+            messages.success(
+                self.request,
+                _("%(setting_type)s updated.")
+                % {
+                    "setting_type": capfirst(self.model._meta.verbose_name),
+                    "instance": self.object,
+                },
+            )
 
-    else:
-        raise NotImplementedError
+        return redirect(
+            "wagtailsettings:edit",
+            app_name=self.model._meta.app_label,
+            model_name=self.model._meta.model_name,
+        )
 
-    return TemplateResponse(
-        request,
-        "wagtailsettings/edit.html",
-        {
-            "opts": model._meta,
-            "setting_type_name": setting_type_name,
-            "instance": instance,
-            "edit_handler": edit_handler,
-            "form": form,
-            "site": site,
-            "site_switcher": site_switcher,
-            "tabbed": isinstance(edit_handler.panel, TabbedInterface),
-            "media": media,
-            "form_id": form_id,
-        },
-    )
+    def form_invalid(self, form):
+        messages.error(self.request, _("The setting could not be saved due to errors."))
+        return super().form_invalid(form)
```

Comment:
This override can probably be removed by setting `error_message = gettext_lazy("The setting could not be saved due to errors.")` on the view class.